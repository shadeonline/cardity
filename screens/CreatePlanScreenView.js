import React, { useState } from 'react';
import ColorSelector from "../components/ColorSelector.js";
import { ScrollView, StyleSheet, Text, View, TextInput, TouchableOpacity } from 'react-native';
import { useNavigation } from '@react-navigation/core';
import { auth, firestore } from '../firebase';
import { collection, addDoc, doc, setDoc } from 'firebase/firestore';





const CreatePlanScreenView = () => {
    const navigation = useNavigation();

    const [description, setDescription] = useState('');
    const [cardName, setCardName] = useState('');
    const [storeName, setStoreName] = useState('');
    const [color, setColor] = useState('#ffc2c7');
    const [rewards, setRewards] = useState([{ stamps: '', reward: '' }]); // Initial state with one empty reward

    // Function to add a new reward field
    const addRewardField = () => {
        setRewards([...rewards, { stamps: '', reward: '' }]);
    };

    const handleCreatePlan = async () => {
        try {
            // Add new loyalty program document
            const loyaltyProgramRef = collection(firestore, 'loyaltyPrograms');
            const newLoyaltyProgram = {
                description: description,
                loyaltyCard: '', // Leave it empty for now
                rewards: parseRewardsInput(), // Parse the rewards input and store as a map
                storeName: storeName,
            };
            const programDocRef = await addDoc(loyaltyProgramRef, newLoyaltyProgram);

            // Add new loyalty card document
            const loyaltyCardRef = collection(firestore, 'loyaltyCards');
            const newLoyaltyCard = {
                cardName: cardName,
                color: color,
                loyaltyProgram: doc(firestore, `loyaltyPrograms/${programDocRef.id}`), // Create a reference here
            };
            const loyaltyCardDocRef = await addDoc(loyaltyCardRef, newLoyaltyCard);

            // Update the loyaltyCard field in newLoyaltyProgram
            const updatedLoyaltyProgram = {
                ...newLoyaltyProgram,
                loyaltyCard: doc(firestore, `loyaltyCards/${loyaltyCardDocRef.id}`), // Create a reference here
            };
            await setDoc(programDocRef, updatedLoyaltyProgram);

            // Navigate back to admin screen or wherever you want
            navigation.goBack();
            console.log("Plan Created!");
        } catch (error) {
            console.error('Error creating plan:', error);
        }
    };

    // Function to parse rewards input and convert it into an map
    const parseRewardsInput = () => {
        const parsedRewards = {};
        rewards.forEach(reward => {
            if (reward.stamps && reward.reward) {
                // Use the stamps as keys and the reward as the value in the map
                parsedRewards[reward.stamps] = reward.reward;
            }
        });
        return parsedRewards;
    };

    return (
        <View style={styles.container}>
            <ScrollView >
                <Text style={styles.heading}>Create Loyalty Plan</Text>

                <Text style={styles.tag}>Card Name:</Text>
                <TextInput
                    style={styles.input}
                    placeholder="i.e: Stag Nite Bar Card"
                    value={cardName}
                    onChangeText={text => setCardName(text)}
                />

                <Text style={styles.tag}>Store Name:</Text>
                <TextInput
                    style={styles.input}
                    placeholder="i.e: Stag Nite Bar"
                    value={storeName}
                    onChangeText={text => setStoreName(text)}
                />

                <Text style={styles.tag}>Loyalty Plan Description:</Text>
                <TextInput
                    style={styles.input}
                    placeholder="i.e: Earn great and awesome reward with this plan"
                    value={description}
                    onChangeText={text => setDescription(text)}
                />

                <Text style={styles.tag}>Color of loyalty card:</Text>
                <ColorSelector onColorChange={setColor} />

                {/* Input fields for rewards */}
                <Text style={styles.tag}>Rewards:</Text>
                {rewards.map((reward, index) => (
                    <View key={index} style={styles.rewardRow}>
                        <TextInput
                            style={styles.rewardInput}
                            placeholder="Stamps required (i.e: 5)"
                            value={reward.stamps}
                            onChangeText={text => {
                                const updatedRewards = [...rewards];
                                updatedRewards[index].stamps = text;
                                setRewards(updatedRewards);
                            }}
                        />
                        <TextInput
                            style={styles.rewardInput}
                            placeholder="Reward (i.e: 1 free drink)"
                            value={reward.reward}
                            onChangeText={text => {
                                const updatedRewards = [...rewards];
                                updatedRewards[index].reward = text;
                                setRewards(updatedRewards);
                            }}
                        />
                    </View>
                ))}

                {/* Add Reward button */}
                <TouchableOpacity style={styles.addRewardButton} onPress={addRewardField}>
                    <Text style={styles.buttonText}>Add Reward</Text>
                </TouchableOpacity>

            </ScrollView>

            <TouchableOpacity style={styles.createPlanButton} onPress={handleCreatePlan}>
                <Text style={styles.buttonText}>Create Plan</Text>
            </TouchableOpacity>
        </View>
    );
};

export default CreatePlanScreenView;

const styles = StyleSheet.create({
    container: {
        flex: 1,
        padding: 20,
    },
    heading: {
        fontSize: 24,
        marginBottom: 20,
    },
    tag: {
        fontSize: 15,
        marginBottom: 5,
    },
    input: {
        marginBottom: 15,
        padding: 10,
        borderWidth: 1,
        borderColor: '#ccc',
        borderRadius: 5,
    },
    createPlanButton: {
        backgroundColor: '#0782F9',
        padding: 15,
        borderRadius: 10,
        alignItems: 'center',
    },
    addRewardButton: {
        backgroundColor: '#14A44D',
        padding: 10,
        borderRadius: 5,
        alignItems: 'center',
        marginTop: 10,
        marginBottom: 60,

    },
    buttonText: {
        color: 'white',
        fontWeight: '700',
        fontSize: 16,
    },
    rewardRow: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginBottom: 5,
    },
    rewardInput: {
        flex: 1,
        marginRight: 5,
        padding: 10,
        borderWidth: 1,
        borderColor: '#ccc',
        borderRadius: 5,
    },
});
