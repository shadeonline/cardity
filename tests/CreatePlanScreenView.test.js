import React from 'react';
import { render } from 'react-native-testing-library';
import CreatePlanScreenView from '../screens/CreatePlanScreenView';
import { NavigationContainer } from '@react-navigation/native';

// Custom mock for useNavigation
const mockNavigation = {
  navigate: jest.fn(),
};

jest.mock('../firebaseFunctions', () => ({
    firebaseFetchProfile: jest.fn(),
    firebaseFetchLoyaltyProgramsByUser: jest.fn(),
  }));

// Mock dependencies
jest.mock('@react-navigation/stack', () => ({
  useNavigation: () => mockNavigation, // Use the custom mock here
}));

jest.mock('../firebase', () => ({
  auth: {
    currentUser: {
      email: 'admin@example.com', // Mock the user's email here
    },
  },
}));

jest.mock('firebase/auth', () => ({
  signOut: jest.fn(), // Mock the signOut function here
}));

jest.mock('../firebaseFunctions', () => ({
  firebaseCreateLoyaltyPlan: jest.fn(), // Mock the function you are testing
}));

describe('CreatePlanScreenView Component', () => {
  it('matches snapshot', () => {
    // Mock data
    const profile = {
      name: 'Admin Name',
    };

    // Render the CreatePlanScreenView component with the mock data
    const { toJSON } = render(
      <NavigationContainer>
        <CreatePlanScreenView profile={profile} />
      </NavigationContainer>
    );

    // Use Jest's expect to match the snapshot
    expect(toJSON()).toMatchSnapshot();
  });
});
