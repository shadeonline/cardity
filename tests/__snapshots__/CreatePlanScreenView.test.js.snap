// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePlanScreenView Component matches snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 20,
    }
  }
>
  <RCTScrollView>
    <View>
      <Text
        style={
          {
            "fontSize": 24,
            "marginBottom": 20,
          }
        }
      >
        Create Loyalty Plan
      </Text>
      <Text
        style={
          {
            "fontSize": 15,
            "marginBottom": 5,
          }
        }
      >
        Card Name:
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="i.e: Stag Nite Bar Card"
        style={
          {
            "borderColor": "#ccc",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginBottom": 15,
            "padding": 10,
          }
        }
        value=""
      />
      <Text
        style={
          {
            "fontSize": 15,
            "marginBottom": 5,
          }
        }
      >
        Store Name:
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="i.e: Stag Nite Bar"
        style={
          {
            "borderColor": "#ccc",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginBottom": 15,
            "padding": 10,
          }
        }
        value=""
      />
      <Text
        style={
          {
            "fontSize": 15,
            "marginBottom": 5,
          }
        }
      >
        Loyalty Plan Description:
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="i.e: Earn great and awesome reward with this plan"
        style={
          {
            "borderColor": "#ccc",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginBottom": 15,
            "padding": 10,
          }
        }
        value=""
      />
      <Text
        style={
          {
            "fontSize": 15,
            "marginBottom": 5,
          }
        }
      >
        Task:
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="i.e: Make a purchase of more than $10"
        style={
          {
            "borderColor": "#ccc",
            "borderRadius": 5,
            "borderWidth": 1,
            "marginBottom": 15,
            "padding": 10,
          }
        }
        value=""
      />
      <Text
        style={
          {
            "fontSize": 15,
            "marginBottom": 5,
          }
        }
      >
        Color of loyalty card:
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 20,
            "marginTop": 5,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#2541b2",
              "borderColor": "green",
              "borderRadius": 20,
              "borderWidth": 5,
              "height": 40,
              "opacity": 1,
              "width": 40,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#1768ac",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 5,
              "height": 40,
              "opacity": 1,
              "width": 40,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#06bee1",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 5,
              "height": 40,
              "opacity": 1,
              "width": 40,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#ffc2c7",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 5,
              "height": 40,
              "opacity": 1,
              "width": 40,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#ffb1f2",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 5,
              "height": 40,
              "opacity": 1,
              "width": 40,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#dd9eff",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 5,
              "height": 40,
              "opacity": 1,
              "width": 40,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#c966ff",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 5,
              "height": 40,
              "opacity": 1,
              "width": 40,
            }
          }
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#792cff",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderWidth": 5,
              "height": 40,
              "opacity": 1,
              "width": 40,
            }
          }
        />
      </View>
      <Text
        style={
          {
            "fontSize": 15,
            "marginBottom": 5,
          }
        }
      >
        Rewards:
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 5,
          }
        }
      >
        <TextInput
          onChangeText={[Function]}
          placeholder="Stamps required (i.e: 5)"
          style={
            {
              "borderColor": "#ccc",
              "borderRadius": 5,
              "borderWidth": 1,
              "flex": 1,
              "marginRight": 5,
              "padding": 10,
            }
          }
          value=""
        />
        <TextInput
          onChangeText={[Function]}
          placeholder="Reward (i.e: 1 free drink)"
          style={
            {
              "borderColor": "#ccc",
              "borderRadius": 5,
              "borderWidth": 1,
              "flex": 1,
              "marginRight": 5,
              "padding": 10,
            }
          }
          value=""
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignSelf": "flex-start",
              "borderRadius": 5,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Text />
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#14A44D",
            "borderRadius": 5,
            "marginBottom": 60,
            "marginTop": 10,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "700",
            }
          }
        >
          Add Reward
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#0782F9",
        "borderRadius": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "700",
        }
      }
    >
      Create Plan
    </Text>
  </View>
</View>
`;
